private_set_intersection/rust/src/client.rs

```diff
@@ -0,0 +1,315 @@
+            let mut null_ptr: *mut c_char = ptr::null_mut();
+            let error_ptr = &mut null_ptr;
+
+            let res_code = psi_client_create_with_new_key(reveal_intersection, &mut client.ctx, error_ptr);
+
+            if res_code != 0 {
+                let error_str = CStr::from_ptr(*error_ptr).to_str().unwrap().to_owned();
+                free(*error_ptr as *mut c_void);
+                return Err(ClientError::new(format!("Failed to create client context: {} ({})", error_str, res_code)));
+            }
+
+            if client.ctx.is_null() {
+                return Err(ClientError::new("Failed to create client context: Context is NULL. This should never happen.".to_string()));
+            }
+
+            Ok(client)
+        }
+    }
+
+    /// Creates a new `PsiClient` instance with the provided private key.
+    ///
+    /// **Warning: this function should be used with caution, since reusing the client key
+    /// for multiple requests can reveal information about the input sets. If in doubt,
+    /// use `PsiClient::create_with_new_key`.**
+    pub fn create_from_key(key: &[u8], reveal_intersection: bool) -> ClientResult<Self> {
+        unsafe {
+            let mut client = Self { ctx: ptr::null_mut(), reveal_intersection: reveal_intersection };
+            let mut null_ptr: *mut c_char = ptr::null_mut();
+            let error_ptr = &mut null_ptr;
+            let key_bytes = PsiClientBuffer {
+                ptr: mem::ManuallyDrop::new(key.to_owned()).as_ptr() as *const c_char,
+                len: key.len() as size_t
+            };
+
+            let res_code = psi_client_create_from_key(key_bytes, reveal_intersection, &mut client.ctx, error_ptr);
+
+            if res_code != 0 {
+                let error_str = CStr::from_ptr(*error_ptr).to_str().unwrap().to_owned();
+                free(*error_ptr as *mut c_void);
+                return Err(ClientError::new(format!("Failed to create client context: {} ({})", error_str, res_code)));
+            }
+
+            if client.ctx.is_null() {
+                return Err(ClientError::new("Failed to create client context: Context is NULL. This should never happen.".to_string()));
+            }
+
+            Ok(client)
+        }
+    }
+
+    /// Creates a request protobuf that can be serialized and sent to the server.
+    ///
+    /// For each input element `x`, this computes `H(x)^c`, where `c` is the client's
+    /// secret key for `ec_cipher`.
+    pub fn create_request<T: AsRef<str>>(&self, raw_input: &[T]) -> ClientResult<Request> {
+        unsafe {
+            let input: Vec<PsiClientBuffer> = raw_input.iter().map(|s| PsiClientBuffer {
+                ptr: s.as_ref().as_ptr() as *const c_char,
+                len: s.as_ref().len() as size_t
+            }).collect();
+
+            let mut null_ptr: *mut c_char = ptr::null_mut();
+            let error_ptr = &mut null_ptr;
+            let mut null_ptr: *mut c_char = ptr::null_mut();
```

Comment:
Is this an intentional shadow of the above? If they are separate for separate uses then maybe they should be named differently so this doesn't get accidentally changed.

Reply:
These are separate instances of `null_ptr`. They serve no other purpose than being valid placeholder pointers that get replaced later on, so there isn't really any risk of them accidentally getting changed. I don't think a different name would be beneficial.

Reply:
You are right, I will rename them.